# CHILDREN'S BOOKS REVIEWS EDA AND DATA CLEANING

## Importing the libraries and data

In [1]:
import pandas as pd
import numpy as np

pd.options.display.max_rows=350
pd.options.display.max_columns=40


import warnings
warnings.filterwarnings('ignore')

In [2]:
filename2 = '../../ALLWOMEN/_MODULE 7/datasets/goodreads_reviews_children.json' #change your path here
data2 = pd.read_json(filename2,lines=True)

In [3]:
data2.shape

(734640, 11)

In [4]:
reviews=data2.copy()

## EDA and Data Cleaning: Dataset of reviews

In [5]:
reviews.head(3)

user_id   book_id  \
0  8842281e1d1347389f2ab93d60773d4d  23310161   
1  8842281e1d1347389f2ab93d60773d4d  17290220   
2  8842281e1d1347389f2ab93d60773d4d   6954929   

                          review_id  rating  \
0  f4b4b050f4be00e9283c92a814af2670       4   
1  22d424a2b0057b18fb6ecf017af7be92       5   
2  50ed4431c451d5677d98dd25ca8ec106       5   

                                         review_text  \
0                        Fun sequel to the original.   
1  One of my favorite books to read to my 5 year ...   
2  One of the best and most imaginative childrens...   

                       date_added                    date_updated  \
0  Tue Nov 17 11:37:35 -0800 2015  Tue Nov 17 11:38:05 -0800 2015   
1  Sat Nov 08 08:54:03 -0800 2014  Wed Jan 25 13:56:12 -0800 2017   
2  Thu Oct 23 13:46:20 -0700 2014  Thu Oct 23 13:47:00 -0700 2014   

                          read_at started_at  n_votes  n_comments  
0                                                   7           0  
1  Tue Jan 24 00:00:00 -0800 2017                   4           0  
2                                                   6           1

In [6]:
reviews.book_id.nunique()
# We have fewer unique book_id in this dataset than in the description dataset

123946

In [7]:
reviews.columns

Index(['user_id', 'book_id', 'review_id', 'rating', 'review_text',
       'date_added', 'date_updated', 'read_at', 'started_at', 'n_votes',
       'n_comments'],
      dtype='object')

Let's drop some columns we don't need in this project

In [8]:
reviews=reviews.drop(['date_added', 'date_updated', 'read_at', 'started_at', 'n_votes',
       'n_comments'], axis=1)

In [9]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 734640 entries, 0 to 734639
Data columns (total 5 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   user_id      734640 non-null  object
 1   book_id      734640 non-null  int64 
 2   review_id    734640 non-null  object
 3   rating       734640 non-null  int64 
 4   review_text  734640 non-null  object
dtypes: int64(2), object(3)
memory usage: 28.0+ MB


In [10]:
reviews.duplicated().sum() 

0

In [12]:
def get_uniques(df,lim=20):
    for col in df.columns:
        array=list(df[col].values)
        uniques=set(array)
        if len(uniques) < lim:
            print(col, ":", uniques)
        else:
            print(col, ":", len(uniques), "unique values.")

In [13]:
get_uniques(reviews)

user_id : 92667 unique values.
book_id : 123946 unique values.
review_id : 734640 unique values.
rating : {0, 1, 2, 3, 4, 5}
review_text : 704137 unique values.


In [14]:
reviews.describe().T

count          mean           std  min       25%        50%  \
book_id  734640.0  9.520406e+06  1.002876e+07  5.0  432463.0  6411331.0   
rating   734640.0  3.818344e+00  1.240084e+00  0.0       3.0        4.0   

                75%         max  
book_id  17349203.0  36469877.0  
rating          5.0         5.0

### User_id

In [15]:
reviews.user_id.value_counts()

a2d6dd1685e5aa0a72c9410f8f55e056    5322
9003d274774f4c47e62f77600b08ac1d    4482
97e2ce2141fa1c880967d78aec3c14fa    3898
6ac35fe952c608da50153d64f616291b    2836
751efd615712748ea54bd36da6e521aa    2824
                                    ... 
168938145a6f059f642bd0eb2b183d5a       1
e4ccf5f545ce85f1eab37856c537b52d       1
54208fdc5f05c626a2765b0d3c480d56       1
e786399c1f21e504609c29808a8d7ec8       1
d699e86772624042c140ac94ca504c91       1
Name: user_id, Length: 92667, dtype: int64

There is an user_id with 5322 different reviews. 

### Rating

In [16]:
reviews.rating.value_counts()

4    253185
5    251400
3    148210
2     40006
0     31113
1     10726
Name: rating, dtype: int64

There are 31113 ratings with a value of 0. We will take a look at this later when we merge the dataframes

### review_id  & review_text

In [17]:
reviews.review_id.nunique()

734640

In [18]:
reviews.review_text.value_counts()

SM                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     2181
A                                                                                                                                                                                                                                                                                                                                                                                                                                           

As we can see, there are 734640 reviews with unique id, but some of the texts of the reviews are repeated.
We will also deal with this after merging the datasets.

## Merging the reviews with the descriptions dataset cleaned before in notebook1

In [20]:
data3=pd.read_csv('descriptionsdfclean.csv') #this is our file from notebook 1

In [21]:
bookstomerge=data3.copy()

In [22]:
bookstomerge.shape

(62395, 15)

In [23]:
bookstomerge.columns

Index(['isbn', 'text_reviews_count', 'is_ebook', 'average_rating',
       'description', 'format', 'publisher', 'num_pages', 'isbn13',
       'publication_year', 'book_id', 'ratings_count', 'title',
       'descriptiondetect', 'titledetect'],
      dtype='object')

In [24]:
reviews.columns

Index(['user_id', 'book_id', 'review_id', 'rating', 'review_text'], dtype='object')

In [25]:
df = pd.merge(reviews,bookstomerge , on=["book_id"])

In [26]:
df.shape

(537462, 19)

In [27]:
df.book_id.nunique() #we have lost 32 books after the merging.

62363

## EDA and Data cleaning of the merged DF

In [28]:
df.isnull().sum().sum()

0

In [29]:
get_uniques(df)

user_id : 73368 unique values.
book_id : 62363 unique values.
review_id : 537462 unique values.
rating : {0, 1, 2, 3, 4, 5}
review_text : 516088 unique values.
isbn : 61037 unique values.
text_reviews_count : 971 unique values.
is_ebook : {False, True}
average_rating : 285 unique values.
description : 62363 unique values.
format : 73 unique values.
publisher : 5366 unique values.
num_pages : 610 unique values.
isbn13 : 62095 unique values.
publication_year : 107 unique values.
ratings_count : 3830 unique values.
title : 62363 unique values.
descriptiondetect : 62363 unique values.
titledetect : 62363 unique values.


### Ratings

In [30]:
df.rating.value_counts()

4    187677
5    183112
3    108046
2     28667
0     22345
1      7615
Name: rating, dtype: int64

In [31]:
df[df.rating==0].nunique() 

user_id                5916
book_id               13067
review_id             22345
rating                    1
review_text           17250
isbn                  12944
text_reviews_count      965
is_ebook                  2
average_rating          237
description           13067
format                   29
publisher              1487
num_pages               448
isbn13                13044
publication_year         86
ratings_count          3175
title                 13067
descriptiondetect     13067
titledetect           13067
dtype: int64

In [32]:
df[df.rating==0]

user_id   book_id  \
9       ff6d428e075e07702736c9d047b2c046  23310161   
57      a33a6e4ca1c8896001a5336fe3bf0ecd  23310161   
61      9cc3f4b9d56415191cfc24f6f59703d9  23310161   
83      4473faba02735ae33825baa557747eab  23310161   
109     5c7f8ab40ef794ea7c5eea62dfe8b8eb  23310161   
...                                  ...       ...   
537347  cd7d78cff86b42120270ba32267dbb8f   3681950   
537354  12458c01d2d31e64784fdea05b4fdd4e  23342191   
537381  453d7864278771acecb3a36ab42d0112  14802030   
537400  4e71cf218924a68104a10318d1e3d0fe   9995871   
537460  b5243a47ce3237056ab824d0bc4ebb87    502386   

                               review_id  rating  \
9       c2cad3a2a5b77fe8b8e2e11d466482f8       0   
57      8bef232bfb6b4e5f6c7d49aad582577d       0   
61      445c2db1565476d1235e86827b433cee       0   
83      1a41bd60925de579955dcc7efee07210       0   
109     36f422be8140cb05472e5969f85698b1       0   
...                                  ...     ...   
537347  c5fddb6598d8f79117269ce09d6ab28a       0   
537354  4cc9ba4a23c92daeb9d6d0f49b13cf0d       0   
537381  4d73092e30ae59d827acf79c4d2f0add       0   
537400  9270a3035d52a4bfc73cbbbc21fa4c29       0   
537460  2dec2121df835c5dd732d4156229ec67       0   

                                              review_text        isbn  \
9       I''m not sure what it is about this kid Duncan...  0399172750   
57      A beautiful, beautiful story, not only for chi...  0399172750   
61      So, so funny! My favorites are Maroon Crayon (...  0399172750   
83      Read some children's books for a short group a...  0399172750   
109                       Another great crayon adventure!  0399172750   
...                                                   ...         ...   
537347  ** spoiler alert ** \n I really enjoyed this b...  0670831964   
537354  "The Boxcar Children guide to adventure: a how...  1497698928   
537381            more of a 13yr olds book.couldnt finish  0141338881   
537400  -two girls \n -penpals \n - two different stor...  1416940227   
537460  See review here: http://pagesofnote.com/babar-...  0810942445   

        text_reviews_count  is_ebook  average_rating  \
9                     1231     False            4.43   
57                    1231     False            4.43   
61                    1231     False            4.43   
83                    1231     False            4.43   
109                   1231     False            4.43   
...                    ...       ...             ...   
537347                   3     False            3.62   
537354                   4      True            4.37   
537381                   3     False            3.90   
537400                   5     False            4.12   
537460                   3     False            3.96   

                                              description     format  \
9       The companion to the #1 blockbuster bestseller...  Hardcover   
57      The companion to the #1 blockbuster bestseller...  Hardcover   
61      The companion to the #1 blockbuster bestseller...  Hardcover   
83      The companion to the #1 blockbuster bestseller...  Hardcover   
109     The companion to the #1 blockbuster bestseller...  Hardcover   
...                                                   ...        ...   
537347  Does a new name make a new you? Which name? Wh...  Hardcover   
537354  The Boxcar Children have long been known for b...      ebook   
537381  Hannah & Joey have been best friends forever. ...  Paperback   
537400  Sincerely, Sophie\nEleven year old Sophie Turn...  Paperback   
537460  Babar's fame has spread to the farthest corner...  Hardcover   

                                       publisher  num_pages         isbn13  \
9                                 Philomel Books       36.0  9780399172755   
57                                Philomel Books       36.0  9780399172755   
61                                Philomel Books       36.0  9780399172755   
83                              

## Language detection of the reviews

In [33]:
def preprocess_df(df):
    
    def process_string(x):
                 
        x = x.lower()
        return x
     
    df['review_text_detect'] = df['review_text'].apply(process_string)
    
    return df

In [34]:
df=preprocess_df(df)

In [35]:
from langdetect import detect

In [36]:
# write the function that detects the language
def language_detection(text):
    try:
        return detect(text)
    except:
        return None

#And neither this code to avoid file overwritting.
df.to_csv('reviews_langdetect.csv', index=False)

After running the language detect function, we have a new csv file. We are importing it below to avoid running the langdetect function code every time we restart the kernel.

### Importing the dataset after applying langdetec 

In [37]:
df=pd.read_csv('reviews_langdetect.csv') #comment this line if you are running the code above

In [38]:
df.shape

(537462, 21)

In [39]:
df.columns

Index(['user_id', 'book_id', 'review_id', 'rating', 'review_text', 'isbn',
       'text_reviews_count', 'is_ebook', 'average_rating', 'description',
       'format', 'publisher', 'num_pages', 'isbn13', 'publication_year',
       'ratings_count', 'title', 'descriptiondetect', 'titledetect',
       'review_text_detect', 'language'],
      dtype='object')

##  2º Round of Data Cleaning

In [49]:
df.language.value_counts()

en    507284
Name: language, dtype: int64

In [41]:
df[df.language!= 'en']

user_id   book_id  \
0       8842281e1d1347389f2ab93d60773d4d  23310161   
39      fdacc154118f88267b3b7bed7aa51080  23310161   
54      2746c2efcd96500931cf35e7a5967175  23310161   
87      c44a8be67a73ef1fab68c8585a60205f  23310161   
89      9d79d624c10936da9a659b45cc2ed2e1  23310161   
...                                  ...       ...   
537427  e84d333d55c67b86fdadfb013150d78c  18224172   
537448  0f777b72a1a35f00f7a948bdfd2dae7d     94752   
537449  0f777b72a1a35f00f7a948bdfd2dae7d   2102369   
537450  0f777b72a1a35f00f7a948bdfd2dae7d    291366   
537460  b5243a47ce3237056ab824d0bc4ebb87    502386   

                               review_id  rating  \
0       f4b4b050f4be00e9283c92a814af2670       4   
39      a9b68ab76c24fbf12f0d3fdfadabaa53       4   
54      f83c80a8c987aea45e42194bba9058bb       4   
87      98179e387c9d60c72c55f7d75e156bbe       5   
89      b5a39cf5352117b7e95ad8501e3b4bc0       4   
...                                  ...     ...   
537427  a876341a2d86af09fbe44fb1d28891bb       5   
537448  e224e25aebe35316375c02b89b48d38f       2   
537449  5741fa4719d0fbbc4d12a186dbc620ef       3   
537450  bb847402b6571b3252ebc2c366a6bea6       3   
537460  2dec2121df835c5dd732d4156229ec67       0   

                                              review_text        isbn  \
0                             Fun sequel to the original.  0399172750   
39                         Yo quiero unos crayones asi:(.  0399172750   
54      (April 7, 2016) https://i.ytimg.com/vi/ZHle5Gr...  0399172750   
87                                  Laugh out loud funny!  0399172750   
89                                               Esteban!  0399172750   
...                                                   ...         ...   
537427                                 I liked this book.  1490591400   
537448           http://bottomshelfbooks.blogspot.com/...  158642114X   
537449           http://bottomshelfbooks.blogspot.com/...  0761452958   
537450           http://bottomshelfbooks.blogspot.com/...  0395922720   
537460  See review here: http://pagesofnote.com/babar-...  0810942445   

        text_reviews_count  is_ebook  average_rating  \
0                     1231     False            4.43   
39                    1231     False            4.43   
54                    1231     False            4.43   
87                    1231     False            4.43   
89                    1231     False            4.43   
...                    ...       ...             ...   
537427                   7     False            3.67   
537448                   5     False            3.94   
537449                   3     False            3.53   
537450                  12     False            4.13   
537460                   3     False            3.96   

                                              description     format  \
0       The companion to the #1 blockbuster bestseller...  Hardcover   
39      The companion to the #1 blockbuster bestseller...  Hardcover   
54      The companion to the #1 blockbuster bestseller...  Hardcover   
87      The companion to the #1 blockbuster bestseller...  Hardcover   
89      The companion to the #1 blockbuster bestseller...  Hardcover   
...                                                   ...        ...   
537427  An adventure for readers aged 199 - 9, stuffed...  Paperback   
537448  Work, Work, Workis the story of an industrious...  Hardcover   
537449  Yoshi the tanuki--a Japanese raccoon-dog--lear...  Hardcover   
537450  On June 14, 1940, Hans and Margret Rey fled Pa...  Hardcover   
537460  Babar's fame has spread to the farthest corner...  Hardcover   

                          publisher  num_pages         isbn13  \
0                    Philomel Books       36.0  9780399172755   
39                   Philomel Books       36.0  9780399172755   
54                   Philomel Books       36.0  9780399172755   
87                   Philomel Books       36.0  9780399172755   
89                 

In [42]:
df[df.language== 'en']

user_id   book_id  \
1       d1e368a7d2870eb6fbf6e0d350568a2d  23310161   
2       7370e3e727f4fdae3ab82f054838d73d  23310161   
3       ec58d8af7f69e19781abed9d25287a57  23310161   
4       cfe4738a5d926fc158d4fcb1287fe2c5  23310161   
5       5a7355dd76fb7649912b783dffe604e1  23310161   
...                                  ...       ...   
537456  12104bd507c4ed850f90b22243905be0  15767746   
537457  0c6a008c66d219aa559aa717d48515a9    139404   
537458  3ad19263e87d15206aa3465aa29d59d9  15263392   
537459  300842609e2359ae76c4fd6ff60704be   3108554   
537461  98d507325b8ab32dab97fb7108c531a2  24941198   

                               review_id  rating  \
1       87fb7685b0b1aa774cc14c330bdcfadf       4   
2       b8c59999e366608b789114d6fdd52a69       5   
3       db7984daa9a8c3b49f4c7f9f73ec86ac       5   
4       f3c100afbe1495fc51bff4d6c3cb7138       4   
5       cde0bf6dfb9ff09ef13494daa0209284       5   
...                                  ...     ...   
537456  8cc43e5a9131e54309e1720dcac1db1e       3   
537457  d7e673822b2d8683be98d7c34b598354       4   
537458  b8cc2beada2f149ee51146282899801d       4   
537459  ce06686588ccd26a4b86f900f15dc0ad       3   
537461  39794ef3c23a2684f3f17ff60cb35f11       5   

                                              review_text        isbn  \
1       Great sequel! Lost crayons write their owner p...  0399172750   
2                           Great sequel to the original.  0399172750   
3               I loved this book as much as the 1st one.  0399172750   
4       It was a cute follow-up to the clever first book.  0399172750   
5       Not as good as the original but fun to have a ...  0399172750   
...                                                   ...         ...   
537456  The Red Gorilla of Oz was alright. I was doing...  147742007X   
537457  This is a great book for creating a home or fo...  0977706303   
537458  Andy Griffith's enough said. The king of child...  0230700756   
537459  A short read about the fear of whats to come w...  019271614X   
537461  simple story, easy words exuding a very powerf...  0143416057   

        text_reviews_count  is_ebook  average_rating  \
1                     1231     False            4.43   
2                     1231     False            4.43   
3                     1231     False            4.43   
4                     1231     False            4.43   
5                     1231     False            4.43   
...                    ...       ...             ...   
537456                   2     False            3.41   
537457                   7     False            3.94   
537458                   5     False            3.92   
537459                   1     False            4.00   
537461                  10     False            4.31   

                                              description       format  \
1       The companion to the #1 blockbuster bestseller...    Hardcover   
2       The companion to the #1 blockbuster bestseller...    Hardcover   
3       The companion to the #1 blockbuster bestseller...    Hardcover   
4       The companion to the #1 blockbuster bestseller...    Hardcover   
5       The companion to the #1 blockbuster bestseller...    Hardcover   
...                                                   ...          ...   
537456  Sebastian is the clumsy, heedless prince of th...    Paperback   
537457  Nominated for the 2007 Anne Izard Storytellers...    Paperback   
537458  Packed with full colour illustrations and diag...    Hardcover   
537459  Although life in London during World War II ha...  not defined   
537461  Poet, novelist, painter, musician and Nobel la...    Paperback   

                         publisher  num_pages         isbn13  \
1                   Philomel Books       36.0  9780399172755   
2                   Philomel Books       36.0  9780399172755   
3                   Philomel Books       36.0  9780399172755   
4                   Philomel Books       36.0  9780399172755   


In [43]:
df=df[df.language== 'en']

In [44]:
df.shape

(507284, 21)

In [45]:
get_uniques(df)

user_id : 68265 unique values.
book_id : 60904 unique values.
review_id : 507284 unique values.
rating : {0, 1, 2, 3, 4, 5}
review_text : 497142 unique values.
isbn : 59599 unique values.
text_reviews_count : 971 unique values.
is_ebook : {False, True}
average_rating : 284 unique values.
description : 60904 unique values.
format : 71 unique values.
publisher : 5267 unique values.
num_pages : 603 unique values.
isbn13 : 60649 unique values.
publication_year : 107 unique values.
ratings_count : 3827 unique values.
title : 60904 unique values.
descriptiondetect : 60904 unique values.
titledetect : 60904 unique values.
review_text_detect : 496634 unique values.
language : {'en'}


In [46]:
df.review_text.value_counts()

O                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            616
Fantastic books for young girls getting into reading!! Great stories about friendship and life lessons. The characters 

In [47]:
df.review_text_detect.value_counts()

o                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  617
fantastic books for young girls getting into reading!! great stories about friendship and life lessons. the characters deal with all sorts of situations and often find responsible solutions to problems. \n i loved this series growing up and wanted to start my own babysitti

In [48]:
df[df.review_text_detect.map(df.review_text_detect.value_counts()) >= 2].sort_values('review_text_detect', ascending=False)

user_id  book_id  \
324853  a1aa241333c125d7149699d2e9a1b037  1306065   
324868  a1aa241333c125d7149699d2e9a1b037   301898   
522452  a1aa241333c125d7149699d2e9a1b037  2288956   
39357   630f1c10f4e76be80e44500189d28ad0     6327   
165671  630f1c10f4e76be80e44500189d28ad0   105999   
...                                  ...      ...   
16745   42ac014557825144b1a71a2e918a8c2c   157993   
17142   78f7aed272cc12087e19ce2bcc20a5de   157993   
220210  6e58a5249317da1274e82a6f97822f13     7779   
220193  f6eb32a7b73ba3e653302623817c9b71     7779   
220197  22fe97031f93e2cc469f115e8692db7a     7779   

                               review_id  rating  \
324853  6cc1da563df43711e48eff3bf1abb801       4   
324868  9e34ef350f8c6fd2d54c2d7a3ea93a5f       4   
522452  872ad62212e2a075da3e6fc4760e0a7d       0   
39357   bb2fdd9729a001d4df6fe9b7132036c5       4   
165671  cb72399dba3f55524885e42079e043ef       3   
...                                  ...     ...   
16745   bb02db6e00c9485670793aa57bff5ba1       5   
17142   9159ce21e7aa717463fa7dc481696ede       4   
220210  3e07ed91948d425ff7b6d153c420e01a       5   
220193  3903b6f117be1780b4681b3da82af06e       5   
220197  f2aa43782c7edbef4fc50a1545010ffb       5   

                                              review_text        isbn  \
324853  {my generic MTH reveiew} \n Miles is now fully...  0375830316   
324868  {my generic MTH reveiew} \n Miles is now fully...  0375830332   
522452  {my generic MTH reveiew} \n Miles is now fully...  0375837280   
39357                           Young Readers Choice 1986  0590032496   
165671                          Young Readers Choice 1986  0380709589   
...                                                   ...         ...   
16745   "And now here is my secret, a very simple secr...  0156012197   
17142   "And now here is my secret, a very simple secr...  0156012197   
220210        "A person's a person, no matter how small."  0679800034   
220193        "A person's a person, no matter how small."  0679800034   
220197        "A person's a person, no matter how small."  0679800034   

        text_reviews_count  is_ebook  average_rating  \
324853                 112     False            3.96   
324868                 142     False            3.99   
522452                   8     False            4.05   
39357                 4302     False            4.17   
165671                1191     False            3.70   
...                    ...       ...             ...   
16745                16639     False            4.28   
17142                16639     False            4.28   
220210                 994     False            4.16   
220193                 994     False            4.16   
220197                 994     False            4.16   

                                              description     format  \
324853  Jack and Annie travel back in time to a desert...  Hardcover   
324868  MERLIN HAS ASKED Jack and Annie to help on ano...  Hardcover   
522452  Merlin the Magician will not eat or sleep or s...  Paperback   
39357   This is not a fairy tale. This is about realwi...  Paperback   
165671  Beverly Cleary's timeless Newbery Medal-winnin...  Paperback   
...                                                   ...        ...   
16745   Moral allegory and spiritual autobiography, Th...  Paperback   
17142   Moral allegory and spiritual autobiography, Th...  Paperback   
220210  Surely among the most lovable of all Dr. Seuss...  Paperback   
220193  Surely among the most lovable of all Dr. Seuss...  Paperback   
220197  Surely among the most lovable of all Dr. Seuss...  Paperback   

                                   publisher  num_pages         isbn13  \
324853  Random House Books for Young Readers      128.0  9780375830310   
324868        Random House for Young Readers      128.0  9780375830334   
522452        Random House for Young Readers      128.0  9780375837289   
39357                        Scholastic Inc.      208.0  978

In [50]:
df=df.sort_values('text_reviews_count', ascending= False)

In [51]:
df.head(4)

user_id  book_id  \
20820  b10fe6b69d79b31aea00153d46fd16c5     3636   
21734  39d0410410b2af134b3dcc66f7a21f5f     3636   
21756  02a401f76c4cb789b41a03c86aba8333     3636   
21755  1d4f47fc8b9c55a2b4b44a32a4243ac3     3636   

                              review_id  rating  \
20820  82574e06eb02f3ccb1cc2423d4276938       5   
21734  da1fadd6ce53c0ab01bcbd1d58c010d9       5   
21756  ff2f2796810b34be15a4cd6adcdf1504       4   
21755  20d31eb9a3c111b19a34b5f551a3bdf5       5   

                                             review_text        isbn  \
20820         SPECTACULAR! This book is simply elegant:)  0385732554   
21734        The best YA science fiction I've ever read!  0385732554   
21756  Before I started reading the Giver I thought I...  0385732554   
21755  I just re-read this book. The first time I rea...  0385732554   

       text_reviews_count  is_ebook  average_rating  \
20820               49850     False            4.12   
21734               49850     False            4.12   
21756               49850     False            4.12   
21755               49850     False            4.12   

                                             description     format publisher  \
20820  Twelve-year-old Jonas lives in a seemingly ide...  Paperback     Ember   
21734  Twelve-year-old Jonas lives in a seemingly ide...  Paperback     Ember   
21756  Twelve-year-old Jonas lives in a seemingly ide...  Paperback     Ember   
21755  Twelve-year-old Jonas lives in a seemingly ide...  Paperback     Ember   

       num_pages         isbn13  publication_year  ratings_count  \
20820      208.0  9780385732550              2006        1311422   
21734      208.0  9780385732550              2006        1311422   
21756      208.0  9780385732550              2006        1311422   
21755      208.0  9780385732550              2006        1311422   

                           title  \
20820  The Giver (The Giver, #1)   
21734  The Giver (The Giver, #1)   
21756  The Giver (The Giver, #1)   
21755  The Giver (The Giver, #1)   

                                       descriptiondetect  \
20820  twelve-year-old jonas lives in a seemingly ide...   
21734  twelve-year-old jonas lives in a seemingly ide...   
21756  twelve-year-old jonas lives in a seemingly ide...   
21755  twelve-year-old jonas lives in a seemingly ide...   

                     titledetect  \
20820  the giver (the giver, #1)   
21734  the giver (the giver, #1)   
21756  the giver (the giver, #1)   
21755  the giver (the giver, #1)   

                                      review_text_detect language  
20820         spectacular! this book is simply elegant:)       en  
21734        the best ya science fiction i've ever read!       en  
21756  before i started reading the giver i thought i...       en  
21755  i just re-read this book. the first time i rea...       en

In [52]:
df1=df.drop_duplicates(subset=['review_text_detect'], keep= 'first')

In [53]:
df1.shape

(496634, 21)

In [54]:
df1.isnull().sum().sum()

0

In [55]:
df1.duplicated().sum()

0

In [56]:
get_uniques(df1)

user_id : 67843 unique values.
book_id : 60004 unique values.
review_id : 496634 unique values.
rating : {0, 1, 2, 3, 4, 5}
review_text : 496634 unique values.
isbn : 58707 unique values.
text_reviews_count : 971 unique values.
is_ebook : {False, True}
average_rating : 284 unique values.
description : 60004 unique values.
format : 70 unique values.
publisher : 5242 unique values.
num_pages : 602 unique values.
isbn13 : 59753 unique values.
publication_year : 106 unique values.
ratings_count : 3824 unique values.
title : 60004 unique values.
descriptiondetect : 60004 unique values.
titledetect : 60004 unique values.
review_text_detect : 496634 unique values.
language : {'en'}


In [57]:
df1.describe().T

count          mean           std     min        25%  \
book_id             496634.0  9.579397e+06  1.002825e+07     5.0  357664.00   
rating              496634.0  3.862239e+00  1.172741e+00     0.0       3.00   
text_reviews_count  496634.0  2.131312e+03  6.835664e+03     1.0      32.00   
average_rating      496634.0  3.984773e+00  3.003256e-01     1.0       3.82   
num_pages           496634.0  1.197532e+02  1.274989e+02     0.0      32.00   
publication_year    496634.0  2.006687e+03  9.890180e+00  1899.0    2003.00   
ratings_count       496634.0  6.965099e+04  2.473229e+05     1.0     154.00   

                           50%          75%         max  
book_id             6609748.00  17380069.00  36444005.0  
rating                    4.00         5.00         5.0  
text_reviews_count      133.00       814.00     49850.0  
average_rating            4.01         4.18         5.0  
num_pages                48.00       195.00      3816.0  
publication_year       2009.00      2013.00      2018.0  
ratings_count           952.00     11951.00   1876252.0

In [58]:
df=df1.drop(['language'],axis=1)

In [59]:
df.shape

(496634, 20)

In [60]:
df.publisher.value_counts().head(50)

HarperCollins                                 27742
HMH Books for Young Readers                   18650
Candlewick Press                              16938
Scholastic Press                              11652
Little, Brown Books for Young Readers         11287
Puffin Books                                  10589
Disney-Hyperion                               10232
Random House Books for Young Readers           9746
Chronicle Books                                8229
Scholastic Inc.                                8003
Atheneum Books for Young Readers               7564
Scholastic                                     7139
Dial Books                                     6804
HarperCollins Publishers                       6479
Greenwillow Books                              6393
Ember                                          5754
Knopf Books for Young Readers                  5528
Simon  Schuster Books for Young Readers        5477
Harry N. Abrams                                5315
Scholastic P

In [61]:
df.publisher.replace({"HarperCollins Publishers": "HarperCollins", 
                      'Scholastic Press':'Scholastic', 'Scholastic Inc.':'Scholastic',
                      'Scholastic, Inc.':'Scholastic', 'Scholastic Paperbacks':'Scholastic','Puffin':'Puffin Books', 
                      'Knopf Books for Young Readers': 'Knopf',
                     'Candlewick Press': 'Candlewick', }, inplace=True)

### Saving the complete clean DF

In [62]:
df.to_csv('FINAL_COMPLETE_DF.csv', index=False)

### And saving a df just with the final book_id to merge later with the descriptions_df

In [63]:
listbook_id=df.book_id.unique()
book_id = pd.DataFrame(listbook_id, columns=["book_id"])

In [64]:
book_id.to_csv('book_id_total.csv', index= False)